### PR TITLE
Addition of tests and code refactor

### PR DESCRIPTION
**Pull request by khalford**
Added tests covering the base feature and github pr module. Refactored the 2 main features into a base feature to be used for future modules. This also reduces code duplication.

Tests Not included:

`read_data.py/validate_required_files()`: This is not an essential function and may be removed
`features/commands.py`: This is going to be reworked in a future PR

### CODE REVIEW

File: cloud-chatops/src/features/base_feature.py [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+
+    def _send_thread_react(self, pr: PrData, channel: str, thread_ts: str) -> None:
+        """
+        This method sends reactions to the PR message if necessary.
+        """
+        react_with = []
+        if pr.old:
+            react_with.append("alarm_clock")
+        if pr.draft:
+            react_with.append("scroll")
```
Comment (DavidFair) [inline cloud-chatops/src/features/base_feature.py:97]: Some bike shedding - I thought `scroll` was comments pending not draft
Could we do something like `building_construction` (a crane) as it's clearer to me at least it's not finished?

File: cloud-chatops/src/read_data.py [hunk 2]
```diff
@@ -4,16 +4,16 @@
...
 import json
-from custom_exceptions import (
+from errors import (
     RepositoriesNotGiven,
     UserMapNotGiven,
     TokensNotGiven,
 )
 
-PATH = "/usr/src/app/cloud_chatops_secrets/"
+PATH = "/usr/src/app/dev_cloud_chatops_secrets/"
```
Comment (DavidFair) [inline cloud-chatops/src/read_data.py:13]: Why is this now explicitly dev? Might be worth a comment explaining how prod vs dev secrets get injected

File: cloud-chatops/tests/test_base_feature.py [hunk 3]
```diff
@@ -0,0 +1,160 @@
+"""Tests for features.base_feature.BaseFeature"""
+
+# pylint: disable=W0212
+# Disabling this as we need to access protected methods to test them
```
Comment (DavidFair) [inline cloud-chatops/tests/test_base_feature.py:4]: If you wrap the methods you expect to be tested this is more representative of how a deriving class will use this?

**Review by DavidFair — CHANGES_REQUESTED** [review]

File: cloud-chatops/src/features/base_feature.py [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+
+    def __init__(self):
+        self.channel = DEFAULT_CHANNEL
+        self.client = WebClient(token=get_token("SLACK_BOT_TOKEN"))
+        self.prs = GetGitHubPRs(get_repos(), DEFAULT_REPO_OWNER).run()
+        self.slack_ids = get_user_map()
+
+    def _github_to_slack_username(self, user: str) -> str:
+        """Makes a call to the global method"""
+        return _github_to_slack_username(self.slack_ids, user)
```
[the remark this answers is not in the document]
Comment (khalford) [reply cloud-chatops/src/features/base_feature.py:39]: Not sure what you mean here

File: cloud-chatops/src/features/base_feature.py (hunk 1 above)
Comment (khalford) [reply cloud-chatops/src/features/base_feature.py:97]: Good idea

File: cloud-chatops/src/features/base_feature.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (anish-mudaraddi) [reply cloud-chatops/src/features/base_feature.py:40]: could you please change the global function name - I kept thinking you're doing some recursive stuff

[the remark this answers is not in the document]
Comment (anish-mudaraddi) [reply cloud-chatops/src/features/base_feature.py:40]: I'd get rid of the global function and just do this:
`self.slack_ids.get(user, None)`

Much easier to understand - plus less code to maintain

**Review by anish-mudaraddi — CHANGES_REQUESTED** [review]
a few things - just some code optimisations

File: cloud-chatops/src/features/base_feature.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (khalford) [reply cloud-chatops/src/features/base_feature.py:40]: Good catch, I knew it could be simpler 😃

**Review by anish-mudaraddi — APPROVED** [review]

**Review by DavidFair — APPROVED** [review]

**Comment by DavidFair** [comment]
Test failures unrelated

**Pull request merged**